docs/source/type_narrowing.rst

```diff
@@ -395,3 +395,157 @@ or rewrite the function to be slightly more verbose:
+        else:
+            reveal_type(x)  # Revealed type is 'int'
+            print(x + 1)  # Valid: x is int
+
+In this example, the function is_str is a type narrowing function
+that returns TypeIs[str]. When used in an if statement, x is narrowed
+to str in the if branch and to int in the else branch.
+
+Key points:
+
+- The function must accept at least one positional argument.
+- The return type is annotated as ``TypeIs[T]``, where ``T`` is the type you
+want to narrow to.
+- The function must return a ``bool`` value.
+- In the ``if`` branch (when the function returns ``True``), the type of the
+ argument is narrowed to the intersection of its original type and ``T``.
+- In the ``else`` branch (when the function returns ``False``), the type of
+the argument is narrowed to the intersection of its original type and the
+complement of ``T``.
+
+TypeIs vs TypeGuard
+-------------------
+
+While both TypeIs and TypeGuard allow you to define custom type narrowing
+functions, they differ in important ways:
+
+- **Type narrowing behavior**: TypeIs narrows the type in both the if and else branches,
+whereas TypeGuard narrows only in the if branch.
+- **Compatibility requirement**: TypeIs requires that the narrowed type T be compatible
+with the input type of the function. TypeGuard does not have this restriction.
+- **Type inference**: With TypeIs, the type checker may infer a more precise type by
+combining existing type information with T.
+
+Here's an example demonstrating the behavior with TypeGuard:
+
+.. code-block:: python
+
+    from typing import TypeGuard, reveal_type
+
+    def is_str(x: object) -> TypeGuard[str]:
+        return isinstance(x, str)
+
+    def process(x: int | str) -> None:
+        if is_str(x):
+            reveal_type(x)  # Revealed type is "builtins.str"
+            print(x.upper())  # ok: x is str
+        else:
+            reveal_type(x)  # Revealed type is "Union[builtins.int, builtins.str]"
+            print(x + 1)  # ERROR: Unsupported operand types for + ("str" and "int")  [operator]
+
+Generic TypeIs
+~~~~~~~~~~~~~~
+
+``TypeIs`` functions can also work with generic types:
+
+.. code-block:: python
+
+    from typing import TypeVar, TypeIs
+
+    T = TypeVar('T')
+
+    def is_two_element_tuple(val: tuple[T, ...]) -> TypeIs[tuple[T, T]]:
+        return len(val) == 2
+
+    def process(names: tuple[str, ...]) -> None:
+        if is_two_element_tuple(names):
+            reveal_type(names)  # Revealed type is 'tuple[str, str]'
+        else:
+            reveal_type(names)  # Revealed type is 'tuple[str, ...]'
+
+
+TypeIs with Additional Parameters
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+TypeIs functions can accept additional parameters beyond the first.
+The type narrowing applies only to the first argument.
+
+.. code-block:: python
+
+    from typing import Any, TypeVar, Type, reveal_type, TypeIs
+
+    T = TypeVar('T')
+
+    def is_instance_of(val: Any, typ: Type[T]) -> TypeIs[T]:
```

Comment:
```suggestion
    from typing import Any, TypeVar, reveal_type, TypeIs

    T = TypeVar('T')

    def is_instance_of(val: Any, typ: type[T]) -> TypeIs[T]:
```
You don't need to use `typing.Type` any more

Reply:
okay! i applied the sugested changes